### PR TITLE
Adding the pfeiffer pressure gauge controller agent.

### PR DESCRIPTION
**Pull request by Zhilei**

### CODE REVIEW

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#script to log and readout pfeiffer TPG 366 gauge contoller
+#via Ethernet connection
+#Zhilei X., Tanay B.
+
+import socket
+import numpy as np
+
+BUFF_SIZE = 128
+ENQ = '\x05'
+IP_ADDRESS = '10.10.10.20'
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: This works for now, but when we add the OCS Agent code we'll want to make this a command line option which will become configuration within the ocs configuration file, that way other devices can be used with the same code either in the same network, or elsewhere.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+class pfeiffer:
+    def __init__(self, ip_address=IP_ADDRESS, port=8000, timeout=10):
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Docstring missing. Please describe the class, its arguments, and attributes.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.ip_address = ip_address
+        self.port = port
+        
+        self.comm = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.comm.connect((self.ip_address, self.port))
+        self.comm.settimeout(timeout)
+    
+    def read_pressure(self, ch_no):
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Docstring formatting should be not as single comments (#), but as a docstring surrounded by sets of triple-quotes ("""). The SO dev guide recommends the Numpy style, example here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy

Please reformat.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        msg = 'PR%d\r\n'%ch_no
+        self.comm.send(msg.encode())
+        status = self.comm.recv(BUFF_SIZE).decode()#Could probably use this to catch exemptions
+        self.comm.send(ENQ.encode())
+        read_str = self.comm.recv(BUFF_SIZE).decode()
+        pressure_str = read_str.split(',')[-1].split('\r')[0]
+        pressure = float(pressure_str)
+        return pressure        
+
+    def read_pressure_all(self):
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Similarly, please reformat the docstring here.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self.comm.send(ENQ.encode())
+        read_str = self.comm.recv(BUFF_SIZE).decode()
+        pressure_str = read_str.split('\r')[0]
+        gauge_states = pressure_str.split(',')[::2]
+        gauge_states = np.array(gauge_states, dtype=int)#this is recorded just incase
+        pressures = pressure_str.split(',')[1::2]
+        pressures = np.array(pressures, dtype=float)
+        return pressures 
+
+    def close(self):
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: While perhaps obvious, given this is one line, a single sentence docstring here too. It provides some description in auto-generated documentation.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 1 above)
Comment (Zhilei) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: I guess the fix will become obvious when the entire agent is witten-up, I will wait to address this later

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 2 above)
Comment (Zhilei) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Done

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 3 above)
Comment (Zhilei) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Fixed

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 4 above)
Comment (Zhilei) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Fixed

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 5 above)
Comment (Zhilei) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Fixed.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 2 above)
Comment (BrianJKoopman) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Formatting is important here for Sphinx to format it correctly, again a good example can be found here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google

The "Attributes" you have listed are the methods of the class, but what is meant by attributes is the variables you've stored for the class, for example, `self.ip_address`, `self.port`, and `self.comm`.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 3 above)
Comment (BrianJKoopman) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Formatting again. I would reformat to:
```
"""Function to measure the pressure of one given channel.

Args:
    ch_no (int): The channel to be measured (e.g. 1-6)

Returns:
    float: The measured pressure

"""
```

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 1 above)
Comment (BrianJKoopman) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: This comment still applies, for others to use this Agent this'll have to not be hardcoded.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
+#script to log and readout pfeiffer TPG 366 gauge contoller
+#via Ethernet connection
+#Zhilei Xu, Tanay Bhandarkar
+
+import socket
+import numpy as np
+from ocs import ocs_agent, site_config, client_t
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: `client_t` is  imported but unused.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    
+    def read_pressure(self, ch_no):
+        """
+        Function to measure the pressure of one given channel
+        ch_no is the chanel to be measured (e.g. 1-6)
+        returns the measured pressure as a float
+        
+        Args:
+            ch_no: The channel to be measured (1-6)
+        Returns:
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Should be a blank line between `Args:` and `Returns:` sections.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            pressure as a float
+        """
+        msg = 'PR%d\r\n'%ch_no
+        self.comm.send(msg.encode())
+        status = self.comm.recv(BUFF_SIZE).decode()#Could probably use this to catch exemptions
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Stale comment?

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        #print(pressure)
+        return pressure        
+
+    def read_pressure_all(self):
+        """measure the pressure of all channel
+        Return an array of 6 pressure values as a float array
+
+        Args:
+            None
+        Returns:
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Similar to above, blank line between sections.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+class pfeifferAgent:
+    def __init__(self, agent,ip_address = IP_ADDRESS, port=8000, f_sample=2.5):
+        self.active = True  
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+        self.f_sample = f_sample
+        self.take_data = False
+        self.module: Optional[Module] = None
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: This looks like it's a line left over from an Agent you based this one on. You never use self.module anywhere, please remove it.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    
+    Attributes:
+       read_pressure reads the pressure from one channel (given as an argument)
+       read_pressure_all reads pressrue from the six channels
+       close closes the socket
+    
+    Todo:
+        -Generalize IP address inputs (input IP  via command line?)
+    """
+    def __init__(self, ip_address=IP_ADDRESS, port=8000, timeout=10, f_sample = 2.5):
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: You should remove the default argument of `IP_ADDRESS` for `ip_address` and expect the IP to be passed in when a pfeiffer object is instantiated in the Agent. (You'll see later, you have an `ip_address` argument for the Agent, but don't actually use it to set the IP address of the device, relying on this default, hard coded value.)

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+class pfeifferAgent:
+    def __init__(self, agent,ip_address = IP_ADDRESS, port=8000, f_sample=2.5):
+        self.active = True  
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+        self.f_sample = f_sample
+        self.take_data = False
+        self.module: Optional[Module] = None
+        self.gauge = pfeiffer()
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Related to an above comment, you're not passing the Agent argument `ip_address` here, when you should be, especially as we adapt to allow a non-hard-coded IP address.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        self.port = port
+        self.ip_address = ip_address
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: You don't use this anywhere, so you don't need a `self.ip_address` attribute, remove this line.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    def __init__(self, agent,ip_address = IP_ADDRESS, port=8000, f_sample=2.5):
+        self.active = True  
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+        self.f_sample = f_sample
+        self.take_data = False
+        self.module: Optional[Module] = None
+        self.gauge = pfeiffer()
+        self.port = port
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Similar to below, you should remove this line if you don't need to keep track of self.port. (This should probably also be passed to the pfeiffer object instantiation.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+        pgroup = parser.add_argument_group('Agent Options')
+        pgroup.add_argument('--ip_address')
+        pgroup.add_argument('--port')
+
+        args = parser.parse_args() 
+
+        site_config.reparse_args(args, 'pfeifferAgent')
+        agent, runner = ocs_agent.init_site_agent(args)
+        pfeiffer_agent = pfeifferAgent(agent)
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: You grab the ip_address and port in argparse, pass them to the pfeifferAgent object here. This will allow users to configure the address and port in their ocs config file, and avoid the hardcoded address issue.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        pgroup.add_argument('--port')
+
+        args = parser.parse_args() 
+
+        site_config.reparse_args(args, 'pfeifferAgent')
+        agent, runner = ocs_agent.init_site_agent(args)
+        pfeiffer_agent = pfeifferAgent(agent) 
+        agent.register_process('acq', pfeiffer_agent.start_acq, pfeiffer_agent.stop_acq, startup = True)
+        agent.register_task('close', pfeiffer_agent.stop_acq)
+        agent.register_task('test', pfeiffer_agent.test_function)
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: This task no longer appears to exist, remove this line.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+            while self.take_data:
+                data = {
+                    'timestamp': time.time(),
+                    'block_name': 'pressures',
+                    'data': {}
+                }
+                pressure_array = self.gauge.read_pressure_all()
+                #Loop through all the channels on the device
+                for i in range(len(pressure_array)):
+                    data['data']["pressure_ch"+str(i)]= pressure_array[i]
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Be aware this is going to index on 0. This is contrary to the convention Bluefors has taken on their vacuum gauge setup.

File: agents/sorenson_dlm/pfeiffer_tpg366_agent.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
+#script to log and readout pfeiffer TPG 366 gauge contoller
```
Comment (BrianJKoopman) [inline agents/sorenson_dlm/pfeiffer_tpg366_agent.py]: This seems like a duplicate of an old copy of the pfeiffer Agent, perhaps meant as a starting point for the DLM agent? If nothing has been done to this to start the DLM agent, please remove it from this PR. The DLM agent should be a separate PR.

File: agents/sorenson_dlm/dlm_agent.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
+#script to log and control the Sorenson DLM power supply, for heaters
```
Comment (BrianJKoopman) [inline agents/sorenson_dlm/dlm_agent.py]: The DLM Agent is a new and different Agent, it should probably be reviewed separately and not committed to this branch.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 17 above)
Comment (tanaybhandarkar) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Are you referring to the name of the channels published? For example, we want to see "Channels 1-6" rather than "Channels 0-5".

Comment (BrianJKoopman) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Yes. Maybe you want to see 0-5, but I wanted to point out the difference since the Bluefors system uses 1-6 on what I believe is an identical gauge system.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py (hunk 14 above)
Comment (BrianJKoopman) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: You still don't need this class attribute, you can remove `self.port = port`.

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 20]
```diff
@@ -1,113 +1,117 @@
...
+from ocs import ocs_agent, site_config
 from ocs.ocs_twisted import TimeoutLock
 import time
 
 BUFF_SIZE = 128
 ENQ = '\x05'
-IP_ADDRESS = '10.10.10.20'
+
 
 class pfeiffer:
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: PEP08 recommendation is classes use CapWords format. Capitalize pfeiffer.

https://www.python.org/dev/peps/pep-0008/#class-names

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 21]
```diff
@@ -1,113 +1,117 @@
...
 
     def close(self):
-        """
-        Close the socket of the connection
-        """
+        """Close the socket of the connection"""
         self.comm.close()
 
+
 class pfeifferAgent:
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Same as the pfeiffer class, please format to `PfeifferAgent`.

https://www.python.org/dev/peps/pep-0008/#class-names

File: agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py [hunk 22]
```diff
@@ -1,113 +1,117 @@
...
+
         Returns:
             pressure as a float
         """
-        msg = 'PR%d\r\n'%ch_no
+        msg = 'PR%d\r\n' % ch_no
         self.comm.send(msg.encode())
-        status = self.comm.recv(BUFF_SIZE).decode()#Could probably use this to catch exemptions
+        # Can use this to catch exemptions, for troubleshooting
+        status = self.comm.recv(BUFF_SIZE).decode()
```
Comment (BrianJKoopman) [inline agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: I just have a question about this. It looks like we're sending a command and reading the result to sort of flush out some sort of internal buffer. I think that because we're not doing anything with the result of that first command. Is this required, or still just a debugging statement?

Comment (tanaybhandarkar) [reply agents/pfeiffer_tpg366/pfeiffer_tpg366_agent.py]: Mainly debugging. It's not necessary, I just didn't want to lose that snippet of code.

**Comment by BrianJKoopman** [comment]
Force pushed after a rebase to catch this branch up to master...

File: docs/agents/pfeiffer.rst [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+block to our ocs configuration file. Here is an example configuration block
+using all of the available arguments::
+
+       {'agent-class': 'PfeifferAgent',
+        'instance-id': 'pfeiffer',
+        'arguments':[
+          ['--ip-address', '10.10.10.20'],
+          ['--port', '8000'],
+          ['--mode', 'acq'],
+          ]
```
Comment (BrianJKoopman) [inline docs/agents/pfeiffer.rst]: You're missing a closing bracket, `}`, here.

File: docs/agents/pfeiffer.rst [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+
+Docker
+``````
+The Pfeiffer Agent can be run via a Docker container. The following is an 
+example of what to insest into your institution's docker-compose file. ::
+
+
+  ocs-pfeiffer:
+    image: ocs-pfeiffer-tpg366-agent:latest
```
Comment (BrianJKoopman) [inline docs/agents/pfeiffer.rst]: This'll be `simonsobs/ocs-pfeiffer-tpg366-agent:latest`. The prefix `simonsobs/` indicates the user on Dockerhub that the image is associated with. Else users will get an error that the image doesn't exist.

File: docs/agents/pfeiffer.rst [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+
+Example Client
+--------------
+Below is an example client to start data acquisition
+
+::
+
+    from ocs.matched_client import MatchedClienti
+    import time
+    pfeiffer = MatchedClient("pfeiffer_1", args=[])
```
Comment (BrianJKoopman) [inline docs/agents/pfeiffer.rst]: This `"pfeiffer_1"` instance-id is inconsistent with the rest of your example, remove the `_1` (or add it in other locations.)

File: docs/agents/pfeiffer.rst [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+
+ocs-config
+``````````
+To configure the Cryomech CPA Agent we need to add a CryomechCPAAgent
+block to our ocs configuration file. Here is an example configuration block
+using all of the available arguments::
+
+       {'agent-class': 'PfeifferAgent',
+        'instance-id': 'pfeiffer',
+        'arguments':[
```
Comment (BrianJKoopman) [inline docs/agents/pfeiffer.rst]: Very nit-picky, but I'd put a space after the `:`.
